README.md
# IPK Project 2

## Network sniffer

## Functionality

The program is a network sniffer that captures packets on a network interface and prints information about them. It uses
the `libpcap` library to capture packets.

It allows users to:

- Specify a network interface to capture packets on.
- Filter packets based on protocol (TCP, UDP, ARP, ICMPv4, ICMPv6, IGMP, MLD).
- Filter packets based on port number for TCP or UDP protocols.
- Specify the number of packets to capture.

The program prints information about each captured packet, including its source and destination IP addresses, protocol,
port number, packet type, and payload.

## Theory

Before diving into the code, it's important to have a basic understanding of the following concepts:

- **Networking protocols:** A protocol is a set of rules that govern the way devices communicate with each other over a
  network. Examples of common networking protocols include TCP, UDP, ARP, ICMPv4, ICMPv6, IGMP, and MLD.
- **Packet headers:** A packet is a unit of data that is transmitted over a network. Each packet has a header that
  contains metadata about the packet, such as the source and destination IP addresses, the protocol used, and the packet
  length.
- **Packet filtering:** Packet filtering is the process of selectively allowing or blocking packets based on certain
  criteria, such as the protocol used, the port number, or the packet type.

In order to extract different types of data from the packets, the program needs to know where to find the relevant
addresses in the packet headers:

- **Time:** The timestamp of the packet is stored in the header of the packet.
- **MAC addresses:** The source and destination MAC addresses are stored in the Ethernet header of the packet.
- **Frame length:** The length of the packet is stored in the header of the packet.
- **IP addresses:** The source and destination IP addresses are stored in the IP header of the packet.
- **Ports:** The source and destination port numbers are stored in the TCP or UDP header of the packet. Note that some
  protocols, such as ARP, ICMPv4, ICMPv6, IGMP, and MLD, do not have port numbers.

## Usage

To compile the program, run the following command:

    make

The program is launched with the following command:

    ./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}

The options are:

- `-i` or `--interface`: specifies the network interface to capture packets on.
- `-p`: specifies the port number to filter packets on.
- `--tcp` or `-t`: filters TCP packets only.
- `--udp` or `-u`: filters UDP packets only.
- `--arp`: filters ARP packets only.
- `--icmp4`: filters ICMPv4 packets only.
- `--icmp6`: filters ICMPv6 echo request/response only.
- `--igmp`: filters IGMP packets only.
- `--mld`: filters MLD packets only.
- `-n`: specifies the number of packets to capture. If not specified, the program will capture one packet. If it is 0,
  the program will capture packets indefinitely.

If there were multiple options specified, the program will output packets that match any of the options. For example,
for

    ./ipk-sniffer -i eth0 -p 80 --tcp --udp --icmp6 -n 10

the program will output packets that match any of the following:

- TCP packets on port 80
- UDP packets on port 80
- ICMPv6 packets

## Testing

The testing was mainly done with the help of Wireshark. Here are some of the tests that were done:

- **Capturing a single TCP packet on port 443 to show that the filters work:**

  Input:

  `./ipk-sniffer -i wlp1s0 --tcp -p 443`

  Expected packet data in the Wireshark:

  ![Wireshark screenshot](./test-img/test-1-tcp.png)

  **Output from the terminal:**

    ```
    timestamp: 2023-04-17T15:40:54.526+02:00
    src mac: 98:77:e7:50:60:5e
    dst mac: 94:08:53:2e:50:cb
    frame length: 66 bytes
    src IP: 140.82.112.21
    dst IP: 192.168.1.49
    src port: 443
    dst port: 53648
    
    0x0000: 94 08 53 2e 50 cb 98 77 e7 50 60 5e 08 00 45 00 ..S.P..w .P`^..E.
    0x0010: 00 34 59 c5 40 00 28 06 3a be 8c 52 70 15 c0 a8 .4Y.@.(. :..Rp...
    0x0020: 01 31 01 bb d1 90 33 bb f3 5d 6b 88 4f 7c 80 10 .1....3. .]k.O|..
    0x0030: 00 aa e2 30 00 00 01 01 08 0a f4 0d e9 ff 78 6e ...0.... ......xn
    0x0040: c9 bb                                           ..
  ```

  Here we can see that the program correctly captured a TCP packet on port 443.
  Everything matches with the packet data from Wireshark screenshot: time, mac addresses, frame length, IP addresses,
  ports and payload data.


- **Capturing an ICMPv6 packet with echo request or response to show different type of output with IPv6 addresses and no
  ports:**

  Input:

  ```./ipk-sniffer -i wlp1s0 --icmp6```

  Expected packet data in the Wireshark:

  ![Wireshark screenshot](./test-img/test-2-icmp6.png)

  **Output from the terminal:**

    ```timestamp: 2023-04-17T16:11:55.099+02:00
    src mac: 98:77:e7:50:60:5e
    dst mac: 33:33:00:00:00:01
    frame length: 118 bytes
    src IP: fe80::9a77:e7ff:fe50:605e
    dst IP: ff02::1
    
    0x0000: 33 33 00 00 00 01 98 77 e7 50 60 5e 86 dd 6c 00 33.....w .P`^..l.
    0x0010: 00 00 00 40 3a ff fe 80 00 00 00 00 00 00 9a 77 ...@:... .......w
    0x0020: e7 ff fe 50 60 5e ff 02 00 00 00 00 00 00 00 00 ...P`^.. ........
    0x0030: 00 00 00 00 00 01 80 00 69 d6 37 03 00 00 00 00 ........ i.7.....
    0x0040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ........ ........
    0x0050: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ........ ........
    0x0060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ........ ........
    0x0070: 00 00 00 00 00 00                               ......
  ```

  Here as well we can see that the program correctly captured an ICMPv6 packet. Everything matches with the packet data
  from the Wireshark screenshot.

- **Wrong inputs to show how my script handles different errors:**

    - _No device specified:_

      `./ipk-sniffer -p 80`

      Will output:

      `Error: no device specified.`
    - _Invalid device specified:_

      `./ipk-sniffer -i eth1000 -p 80`

      Will output:

      `Error: device eth1000 does not exist.`
    - _Invalid port number:_

      `./ipk-sniffer -i eth0 -p 100000`

      Will output:

      `Error: invalid port number: 100000`

## Sources

Here are some sources that were helpful for understanding `libpcap`:

- [tcpdump tutorial](https://danielmiessler.com/study/tcpdump/)
- [An introduction to libpcap](https://www.tcpdump.org/pcap.html)
- [Using libpcap in C](https://www.devdungeon.com/content/using-libpcap-c)
- [libpcap documentation](https://www.tcpdump.org/manpages/pcap.3pcap.html)
- [pcap-filter man page](https://www.tcpdump.org/manpages/pcap-filter.7.html)
- [lectures from IPK 2022/2023](https://moodle.vut.cz/mod/folder/view.php?id=289124)

CHANGELOG.md
## Implemented functions

Main function:

    int main(int argc, char **argv)

In the main function, we initialize the filters, parse arguments, and open device for capturing packets.

To parse arguments, we have a function called `parse_args`:

    void parse_args(int argc, char **argv)

It uses `getopt_long` function and `option` struct to parse arguments and set the global flags and variables. Here we
also check if the input given by user was valid.

To create the filter we have a function called `set_filter`:

    void set_filter(char *filter_exp)

Depending on the flags set by user, we create a filter expression and set the filter. To add filters we
use `ADD_TO_FILTER_EXPRESSION` macros, since we have to check if the filter expression is empty or not, and if we need
to add `"or"` to the beginning of a new filter.
Then we use `pcap_compile` function to compile the filter expression and `pcap_setfilter` function to set the filter.

Then we start capturing packets with the `pcap_loop` function. It uses our function `packet_handler` as a callback
function:

    void packet_handler(u_char *args, const struct pcap_pkthdr *header, const u_char *packet)

It shows all the packet data, and for each type of data we have a function:

- `print_timestamp(struct timeval time_in_tv)` - prints out the time of the packet.
- `print_mac_addresses(uint8_t *ether_shost, uint8_t *ether_dhost)` - prints out the source and destination MAC.
  addresses
- `print_ip_addresses(u_char *packet, uint16_t ether_type)` - prints out the source and destination IP addresses. For
  different protocols IP addresses are stored in different places in the packet header, so we have to check the protocol
  type.
- `print_ports(const u_char *packet, uint16_t ether_type)` - prints out the source and destination port numbers. For
  TCP, UDP and ARP
  protocols port numbers are stored in different places in the packet header, so we have to check the protocol type.
  Every other protocol does not have port numbers, so we don't print anything.
- `print_packet_data(const u_char *packet, uint32_t size)` - prints out the payload of the packet.

## Used libraries

The program uses following libraries:

```
stdio.h
stdlib.h
string.h
unistd.h
sys/types.h
sys/socket.h
netinet/in.h
arpa/inet.h
pcap.h
time.h
net/ethernet.h
getopt.h
```

ipk-sniffer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <pcap.h>
#include <time.h>
#include <net/ethernet.h>
#include <getopt.h>

// macros to add expression to filter
// @param expression - expression to add
// @param filter - filter to add expression to
// @param ignore_or - flag to ignore "or"
#define ADD_TO_FILTER_EXPRESSION(expression, filter, ignore_or) \
        if (strlen(filter) > 0) { \
            if (!ignore_or) { \
            strcat(filter, " or "); \
            } \
        } \
        strcat(filter, expression); \

//flags
int tcp = 0, // flag for tcp packets - will only show tcp packets
    udp = 0, // flag for udp packets - will only show udp packets
    arp = 0, // flag for arp frames
    icmp4 = 0, // flag for ICMPv4 packets
    icmp6 = 0, // flag for ICMPv6 echo request/response
    ndp = 0, // flag for ICMPv6 NDP packets
    igmp = 0, // flag for IGMP packets
    mld = 0, // flag for mld packets
    packets_count = 1, // number of packets to capture - if not specified equals to 1
    port = -1;  // port to filter on

char *device = NULL; // device to capture on

// device validation
// @param device_name - name of the device
void validate_device(char *device_name) {
    pcap_if_t *alldevs;
    char errbuf[PCAP_ERRBUF_SIZE];
    if (pcap_findalldevs(&alldevs, errbuf) == -1) {
        fprintf(stderr, "Error finding devices: %s\n", errbuf);
        exit(EXIT_FAILURE);
    }
    pcap_if_t *d;
    for (d = alldevs; d; d = d->next) {
        if (strcmp(d->name, device_name) == 0) {
            pcap_freealldevs(alldevs);
            return;
        }
    }
    pcap_freealldevs(alldevs);
    fprintf(stderr, "Error: device %s does not exist.\n", device_name);
    exit(EXIT_FAILURE);
}

// port validation
void validate_port() {
    if (port < 0 || port > 65535) {
        fprintf(stderr, "Error: invalid port number: %d\n", port);
        exit(EXIT_FAILURE);
    }
}

// print all available devices
void print_devices() {
    printf("\n");
    pcap_if_t *alldevs, *d;
    char errbuf[PCAP_ERRBUF_SIZE];
    if (pcap_findalldevs(&alldevs, errbuf) == -1) {
        fprintf(stderr, "Error finding devices: %s\n", errbuf);
        exit(EXIT_FAILURE);
    }
    for (d = alldevs; d; d = d->next) {
        printf("%s\n", d->name);
    }
    printf("\n");
    pcap_freealldevs(alldevs);
}

// parse command line arguments
// @param argc - number of arguments
// @param argv - array of arguments
void parse_args(int argc, char **argv) {
    // exit if no arguments were given
    if (argc == 1) {
        print_devices();
        exit(0);
    }

    struct option long_options[] = {
            {"interface", optional_argument, NULL, 'i'},
            {"tcp",       no_argument, &tcp,       1},
            {"udp",       no_argument, &udp,       1},
            {"arp",       no_argument, &arp,       1},
            {"icmp4",     no_argument, &icmp4,     1},
            {"icmp6",     no_argument, &icmp6,     1},
            {"igmp",      no_argument, &igmp,      1},
            {"mld",       no_argument, &mld,       1},
            {NULL, 0,                        NULL, 0}
    };
    int opt, option_index;

    while ((opt = getopt_long(argc, argv, "i::p:tun:", long_options, &option_index)) != -1) {
        switch (opt) {
            case 'i':
                // if -i is given without argument, print all available devices and exit
                if (argv[optind] == NULL) {
                    if (argc == 2) {
                        print_devices();
                        exit(0);
                    } else {
                        fprintf(stderr, "Error: no device specified.\n");
                        exit(EXIT_FAILURE);
                    }
                }
                device = argv[optind];
                validate_device(device);
                break;
            case 'p':
                port = atoi(optarg);
                validate_port();
                break;
            case 't':
                tcp = 1;
                break;
            case 'u':
                udp = 1;
                break;
            case 'n':
                packets_count = atoi(optarg);
                break;
            case '?':
                exit(EXIT_FAILURE);
            default:
                break;
        }
    }

    if (device == NULL) {
        fprintf(stderr, "Error: no device specified.\n");
        exit(EXIT_FAILURE);
    }
}

// convert time from timeval to string in format: YYYY-MM-DDTHH:MM:SS.MICROSEC+HH:MM
// @param time_in_tv - time in timeval format
void print_timestamp(struct timeval time_in_tv) {
    char buffer[128];
    char time_buffer[64];
    time_t time_in_sec = time_in_tv.tv_sec;
    struct tm *time = localtime(&time_in_sec);

    strftime(time_buffer, sizeof time_buffer, "%FT%T", time);
    snprintf(buffer, sizeof buffer, "%s.%03ld", time_buffer, time_in_tv.tv_usec / 1000);

// add timezone in HH:MM format
    char timezone_buffer[6];
    char timezone_hours[2];
    char timezone_minutes[2];

    strftime(timezone_buffer, 10, "%z", time);
    strncpy(timezone_hours, timezone_buffer + 1, 2);
    strncpy(timezone_minutes, timezone_buffer + 3, 2);

    strncat(buffer, "+", 2);
    strncat(buffer, timezone_hours, 2);
    strncat(buffer, ":", 2);
    strncat(buffer, timezone_minutes, 2);

    printf("timestamp: %s\n", buffer);
}

void print_mac_addresses(uint8_t *ether_shost, uint8_t *ether_dhost) {
    printf("src mac: %02x:%02x:%02x:%02x:%02x:%02x\n",
           ether_shost[0], ether_shost[1], ether_shost[2],
           ether_shost[3], ether_shost[4], ether_shost[5]);
    printf("dst mac: %02x:%02x:%02x:%02x:%02x:%02x\n",
           ether_dhost[0], ether_dhost[1], ether_dhost[2],
           ether_dhost[3], ether_dhost[4], ether_dhost[5]);
}

// print IP addresses
// @param packet - pointer to the beginning of the packet
// @param ether_type - ether type of the packet
void print_ip_addresses(u_char *packet, uint16_t ether_type) {
    char src_ip[INET6_ADDRSTRLEN];
    char dst_ip[INET6_ADDRSTRLEN];

    uint16_t protocol = ntohs(ether_type);
    // find IP addresses depending on the ether type
    if (protocol == ETHERTYPE_IP) {
        inet_ntop(AF_INET, packet + 26, src_ip, INET_ADDRSTRLEN);   // ip addresses on 26 and 30 bytes of the packet
        inet_ntop(AF_INET, packet + 30, dst_ip, INET_ADDRSTRLEN);
    } else if (protocol == ETHERTYPE_IPV6) {
        inet_ntop(AF_INET6, packet + 22, src_ip, INET6_ADDRSTRLEN); // ip addresses on 22 and 38 bytes of the packet
        inet_ntop(AF_INET6, packet + 38, dst_ip, INET6_ADDRSTRLEN);
    } else if (protocol == ETHERTYPE_ARP) {
        inet_ntop(AF_INET, packet + 28, src_ip, INET_ADDRSTRLEN);   // ip addresses on 28 and 38 bytes of the packet
        inet_ntop(AF_INET, packet + 38, dst_ip, INET_ADDRSTRLEN);
    }

    if (protocol == ETHERTYPE_IP || protocol == ETHERTYPE_IPV6 || protocol == ETHERTYPE_ARP) {
        printf("src IP: %s\n", src_ip);
        printf("dst IP: %s\n", dst_ip);
    }
}

// print ports
// @param packet - pointer to the beginning of the packet
// @param ether_type - ether type of the packet
void print_ports(const u_char *packet, uint16_t ether_type) {
    char src_port[6];
    char dst_port[6];

    uint16_t protocol = ntohs(ether_type);
    // find ports depending on the ether type
    if (protocol == ETHERTYPE_IP) {
        // print ports only if IP protocol is TCP or UDP
        if (packet[23] == IPPROTO_TCP || packet[23] == IPPROTO_UDP) {   // byte 23 is IPv4 protocol
            snprintf(src_port, 6, "%d", ntohs(*(uint16_t *) (packet + 34)));    // ports on bytes 34 and 36
            snprintf(dst_port, 6, "%d", ntohs(*(uint16_t *) (packet + 36)));

            printf("src port: %s\n", src_port);
            printf("dst port: %s\n", dst_port);
        }
    } else if (protocol == ETHERTYPE_IPV6) {
        // print ports only if IP protocol is TCP or UDP
        if (packet[20] == IPPROTO_TCP || packet[20] == IPPROTO_UDP) {   // byte 20 is IPv6 protocol
            snprintf(src_port, 6, "%d", ntohs(*(uint16_t *) (packet + 54)));    // ports on bytes 54 and 56
            snprintf(dst_port, 6, "%d", ntohs(*(uint16_t *) (packet + 56)));

            printf("src port: %s\n", src_port);
            printf("dst port: %s\n", dst_port);
        }
    }
}


// print packet data in hex and ascii
// @param packet - pointer to packet data
// @param size - size of packet data
void print_packet_data(const u_char *packet, uint32_t size) {
    printf("\n");

    int symbols_printed = 0;    // number of symbols printed in current line to help print data in the last line

    for (int i = 0; i < size; i++) {
        // print address in hex
        if (i % 16 == 0) {
            printf("0x%04x: ", i);
            symbols_printed = 0;
        }
        symbols_printed++;

        printf("%02x ", packet[i]); // print packet data in hex

        if (i % 16 == 15) {
            // print packet data in ascii
            for (int j = i - 15; j <= i; j++) {
                // spaces after every 8 symbols
                if (j % 8 == 0 && j != i - 15) {
                    printf(" ");
                }

                // print only printable characters
                if (packet[j] >= 32 && packet[j] <= 126) {
                    printf("%c", packet[j]);
                } else {
                    printf(".");
                }
            }
            printf("\n");
        }
    }

    // print packet data in ascii for last line
    if (symbols_printed % 16 != 0) {
        for (int i = 0; i < 16 - symbols_printed; i++) {
            printf("   ");
        }
        for (uint32_t i = size - symbols_printed; i < size; i++) {
            if (i % 8 == 0 && i != size - symbols_printed) {
                printf(" ");
            }
            if (packet[i] >= 32 && packet[i] <= 126) {
                printf("%c", packet[i]);
            } else {
                printf(".");
            }
        }
    }

    printf("\n\n");
}

// callback function for pcap_loop
// @param args - user supplied argument
// @param header - pointer to the pcap_pkthdr structure
// @param packet - pointer to the packet data
void packet_handler(u_char *args, const struct pcap_pkthdr *header, const u_char *packet) {
//  get packet header
    struct ether_header *eth_header;
    eth_header = (struct ether_header *) packet;

//  print timestamp
    print_timestamp(header->ts);

//  print source and destination mac addresses
    print_mac_addresses(eth_header->ether_shost, eth_header->ether_dhost);

//    print frame length
    printf("frame length: %d bytes\n", header->len);

//  print src and dst IP addresses
    print_ip_addresses((u_char *) packet, eth_header->ether_type);

//  print ports
    print_ports((u_char *) packet, eth_header->ether_type);

//  print payload data
    print_packet_data(packet, header->len);
}

// create a filter expression for pcap_setfilter
// @param filter_exp - filter expression
void set_filter(char *filter_exp) {
    // add port to filter expression
    if (port != -1) {
        char port_str_buf[16];
        snprintf(port_str_buf, 16, "port %d", port);
        ADD_TO_FILTER_EXPRESSION(port_str_buf, filter_exp, 0)

        // if both tcp and udp are set or both are unset, get both to filter
        if (tcp && udp || !tcp && !udp) {
            ADD_TO_FILTER_EXPRESSION(" and (tcp or udp)", filter_exp, 1)
        }
            // if only tcp or udp is set, get only that protocol
        else if (tcp) {
            ADD_TO_FILTER_EXPRESSION(" and tcp", filter_exp, 1)
        } else if (udp) {
            ADD_TO_FILTER_EXPRESSION(" and udp", filter_exp, 1)
        }
    } else {
        // if port is not set just get tcp or udp
        if (tcp) {
            ADD_TO_FILTER_EXPRESSION("tcp", filter_exp, 0)
        }
        if (udp) {
            ADD_TO_FILTER_EXPRESSION("udp", filter_exp, 0)
        }
    }

    // add icmpv4 to filter expression
    if (icmp4) {
        ADD_TO_FILTER_EXPRESSION("icmp", filter_exp, 0)
    }

    // add icmpv6 echo request/response to filter expression
    if (icmp6) {
        ADD_TO_FILTER_EXPRESSION("icmp6 and (icmp6[0] == 128 or icmp6[0] == 129)", filter_exp, 0)
    }

    // add arp to filter expression
    if (arp) {
        ADD_TO_FILTER_EXPRESSION("arp", filter_exp, 0)
    }

    // add ndp to filter expression
    if (ndp) {
        ADD_TO_FILTER_EXPRESSION(
                "icmp6 and (icmp6[0] == 133 or icmp6[0] == 134 or icmp6[0] == 135 or icmp6[0] == 136 or icmp6[0] == 137 or icmp6[0] == 148 or icmp6[0] == 149)",
                filter_exp, 0)
    }

    if (igmp) {
        ADD_TO_FILTER_EXPRESSION("igmp", filter_exp, 0)
    }

    if (mld) {
        ADD_TO_FILTER_EXPRESSION(
                "icmp6 and (icmp6[0] == 130 or icmp6[0] == 131 or icmp6[0] == 132 or icmp6[0] == 143)",
                filter_exp, 0)
    }
}


int main(int argc, char **argv) {
//    inits
    char errbuf[PCAP_ERRBUF_SIZE];
    int timeout_limit = 1500; /* In milliseconds */
    bpf_u_int32 subnet_mask, ip;
    struct bpf_program filter;
    char filter_exp[256] = ""; // filter expression

//    parse arguments and get all the flags
    parse_args(argc, argv);

//    create filter expression
    set_filter(filter_exp);

    pcap_t *handle = pcap_open_live(device, BUFSIZ, 1, 1000, errbuf);
    if (handle == NULL) {
        fprintf(stderr, "Error opening device %s: %s\n", device, errbuf);
        exit(EXIT_FAILURE);
    }

    if (pcap_lookupnet(device, &ip, &subnet_mask, errbuf) == -1) {
        printf("Could not get information for device: %s\n", device);
        ip = 0;
        subnet_mask = 0;
    }

    handle = pcap_open_live(device, 1028, 1, timeout_limit, errbuf);
    if (handle == NULL) {
        fprintf(stderr, "Could not open device %s: %s\n", device, errbuf);
        return 2;
    }

    if (pcap_datalink(handle) != DLT_EN10MB) {
        fprintf(stderr, "Device %s doesn't provide Ethernet headers - not supported\n", device);
        return (2);
    }

    if (pcap_compile(handle, &filter, filter_exp, 0, ip) == -1) {
        printf("Bad filter - %s\n", pcap_geterr(handle));
        return 2;
    }

    if (pcap_setfilter(handle, &filter) == -1) {
        printf("Error setting filter - %s\n", pcap_geterr(handle));
        return 2;
    }

//    capture packets
    pcap_loop(handle, packets_count, packet_handler, NULL);

//    close the handle
    pcap_close(handle);

    return 0;
}